components/tikv_util/src/resizable_threadpool.rs

```diff
@@ -1,161 +1,359 @@
 // Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
 
-use std::sync::{Arc, RwLock};
+use std::{
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc, Mutex, Weak,
+    },
+    time::Duration,
+};
 
 use futures::Future;
-use tokio::{io::Result as TokioResult, runtime::Runtime};
-
-/// DaemonRuntime is a "background" runtime, which contains "daemon" tasks:
-/// any task spawn into it would run until finish even the runtime isn't
-/// referenced.
-pub struct DaemonRuntime(Option<Runtime>);
-
-impl DaemonRuntime {
-    /// spawn a daemon task to the runtime.
-    pub fn spawn(self: &Arc<Self>, f: impl Future<Output = ()> + Send + 'static) {
-        let wkr = self.clone();
-        self.0.as_ref().unwrap().spawn(async move {
-            f.await;
-            drop(wkr)
-        });
-    }
+use tokio::{
+    io::Result as TokioResult,
+    runtime::{Builder, Runtime},
+    time::interval,
+};
 
-    /// create a daemon runtime from some runtime.
-    pub fn from_runtime(rt: Runtime) -> Arc<Self> {
-        Arc::new(Self(Some(rt)))
-    }
+struct DeamonRuntime {
+    inner: Option<Runtime>,
+    task_count: Arc<AtomicUsize>,
 }
 
-impl Drop for DaemonRuntime {
+impl Drop for DeamonRuntime {
     fn drop(&mut self) {
-        if let Some(runtime) = self.0.take() {
+        if let Some(runtime) = self.inner.take() {
             runtime.shutdown_background();
         }
     }
 }
 
+#[derive(Clone)]
+pub struct DeamonRuntimeHandle {
+    inner: Weak<DeamonRuntime>,
+}
+
+impl DeamonRuntimeHandle {
+    pub fn spawn<Fut>(&self, fut: Fut)
+    where
+        Fut: Future<Output = ()> + Send + 'static,
+        {
+            let rc_runtime = match self.inner.upgrade() {
+                Some(runtime) => runtime,
+                None => return,
+            };
+        
+            let inner = match rc_runtime.inner.as_ref() {
+                Some(inner) => inner,
+                None => return,
+            };
+        
+            let handle = inner.handle().clone();
+            let task_count = rc_runtime.task_count.clone();
+        
+            task_count.fetch_add(1, Ordering::SeqCst);
+            
+            handle.spawn(async move {
```

Comment:
you don't need to take the `handle` to call `spawn` or `block_on`.

```suggestion
            inner.spawn(async move {
```